### PR TITLE
Add a blog post on Go and Rust

### PR DESCRIPTION
**Pull request by nrc**
PTAL @Wenting0905 @CaitinChen @dcalvin

### CODE REVIEW

File: go-from-rust.md [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+It is a lot less frustrating and a lot more productive than using C/C++, Java, Python, etc. It did feel, however, like part of that generation of languages. It has learnt some lessons from them, and I think it is probably the best language of that generation; but it is definitely part of that generation. It is an incremental improvement, rather than doing something radically different (to be clear, this is not a value judgement, incremental is often good in software engineering). A good example of this is `nil`: languages like Rust and Swift have removed the concept of `null` and eliminated a whole class of errors. Go renames it to `nil` (why? No idea) and makes it a bit less dangerous (no null *values*, distinguishes between `nil` and `0`). But the core idea is still there, and so is the common runtime error of dereferencing a null pointer.
```
Comment (gregwebs) [inline go-from-rust.md]: Ruby uses nil and I believe it came from Lisp, so I wouldn't say they are renaming anything, just using a name that does already have a tradition and mean the same thing.

File: go-from-rust.md [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+* Having a garbage collector! Not thinking about memory management really does make programming easier.
+* Varargs
+
+## Things I did not like
+
+In no particular order.
+
+* `nil` slices - `nil`,  a `nil` slice, and an empty slice are all different. I'm pretty sure you only need two of those things, not three.
+* No first class enums. Using constants feels backwards.
+* Disallowing import cycles. This really limits how useful packages are for modularising a project, since it encourages putting lots of files in a package.
```
Comment (gregwebs) [inline go-from-rust.md]: That's the approach usually taken by go users, but I actually solve it by using lots of really small packages. Whenever there is an import cycle that just means there should be another package.

File: go-from-rust.md [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+* `switch` may be non-exhaustive.
+* `for ... range` returns a pair of index/value. Getting just the index is easy (just ignore the value) but getting just the value requires being explicit. This seems back-to-front to me since I need the value and not the index in most cases.
+* Syntax:
+  - inconsistency between definitions and uses.
+  - Pickiness of the compiler (e.g., requiring or forbidding trailing commas); this is somewhat alleviated by good tooling.
```
Comment (gregwebs) [inline go-from-rust.md]: I would say it is entirely eliminated by good tooling (VSCode running gofmt).

File: go-from-rust.md [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+## Things I like
+
+* Compile times. Definitely quick; definitely a lot quicker than Rust. But not actually as quick as I was expecting (it feels to me to be similar to C/C++ for medium to large projects, I was kind of expecting instant).
+* Go routines and channels. Having lightweight syntax for spawning Go routines and using channels is really nice. It really shows the power of syntax that such a small detail makes concurrent programming feel so much nicer than in other languages.
+* Interfaces. They're not very sophisticated but they are easy to understand and use, and useful in a lot of places.
+* `if ...; ... { }` syntax. Being able to restrict the scope of variables to the body of `if` statements is nice. This has a similar effect to `if let` in Swift and Rust, but is more general purpose (Go does not have pattern matching like Swift and Rust, so it cannot use `if let`.
+* Testing and doc comments are easy to use.
+* The `Go` tool is nice - having everything in one place rather than requiring multiple tools on the command line.
+* Having a garbage collector! Not thinking about memory management really does make programming easier.
```
Comment (CaitinChen) [inline go-from-rust.md]:
```suggestion
* Having a garbage collector (GC)! Not thinking about memory management really does make programming easier.
```

File: go-from-rust.md [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+date: 2020-02-19
+summary:  Read this post to learn more. 
+tags: ['Engineering']
+categories: ['Engineering']
+image: 
+---
+
+# Early Impressions of Go from a Rust Programmer
+
+I've been using Go for the past few weeks. It's my first time using Go for a large (-ish), serious project. I've previously looked at Go a lot and played with examples and toy programs when researching features for Rust. Real programming is a very different experience.
```
Comment (CaitinChen) [inline go-from-rust.md]:
```suggestion
I've been using [Go](https://en.wikipedia.org/wiki/Go_(programming_language)) for the past few weeks. It's my first time using Go for a large (-ish), serious project. I've previously looked at Go a lot and played with examples and toy programs when researching features for [Rust](https://en.wikipedia.org/wiki/Rust_(programming_language)). Real programming is a very different experience.
```

File: go-from-rust.md [hunk 6]
```diff
@@ -0,0 +1,89 @@
+---
+title: Early Impressions of Go from a Rust Programmer
+author: ['Nick Cameron']
+date: 2020-02-19
+summary:  Read this post to learn more.
```
Comment (CaitinChen) [inline go-from-rust.md]:
```suggestion
summary: Nick Cameron is a Rust programmer who has been using Go for a few weeks. In this post, he talks about his early impressions of Go. Read this post to learn more. 
```

**Pull request merged**